src/lcd_st7789v.c

```diff
@@ -0,0 +1,188 @@
+// Commands for sending messages to an st7789V display connected
+// using the "8080-II series" interface
+//
+// Copyright (C) 2020  Martijn van Buul <martijn.van.buul@gmail.com>
+// Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include "autoconf.h" // CONFIG_CLOCK_FREQ
+#include "basecmd.h" // oid_alloc
+#include "board/gpio.h" // gpio_out_write
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // timer_from_us
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // DECL_SHUTDOWN
+
+struct st7789v {
+    uint8_t last_dcx;
+    uint8_t last_data;
+    struct gpio_out dcx, wrx, databits[8];
+};
+
+
+/****************************************************************
+ * Transmit functions
+ ****************************************************************/
+
+// Write 8 bits to the st7789v using the "8080-II series" interface
+static inline void
+st7789v_xmit_byte(struct st7789v *h, uint8_t data, uint8_t dcx )
+{
+    uint8_t mask = 0x01;
+    gpio_out_write(h->wrx, 0);
+    if (h->last_dcx != dcx)
+    {
+        gpio_out_write(h->dcx, dcx);
+        h->last_dcx = dcx;
+    }
+
+    for (int bit = 0; bit < 8; ++bit, mask <<=1)
+    {
+        const uint8_t new_data = data & mask;
+        if ( new_data != (h->last_data & mask) )
+        {
+            gpio_out_write( h->databits[bit], new_data);
+        }
+    }
```

Comment:
FYI, comparing to a previous value is a pessimization on most micro-controllers. It's faster to just unconditionally write the desired value.

What micro-controller are you currently using for the display?  (I know the original used an stm32f4, but I got the impression you wired it to some other micro-controller.)

Reply:
I'm using a Blue Pill (STM32F106). gpio_out_write() isn't a simple write though, it is much more complicated (although a lot of it gets inlined apparently). I originally did *not* skip the write, but some probing with a logic analyzer suggested that this was worthwhile. I lost that particular trace, but it's easily repeated.

Reply:
The compiler should be able to turn the conditional into an optimized shift in src/stm32/gpio.c .  I'd be surprised if the memory load and actual test-and-branch would be faster.

Reply:
> The compiler should be able to turn the conditional into an optimized shift in src/stm32/gpio.c . I'd be surprised if the memory load and actual test-and-branch would be faster.

I finally got around to measuring this. I apologize for the thumbnail size of the bitmaps; my oscilloscope is showing its age. In all images, I'm probing RDX (yellow) and D12 (blue). WRX is pulsed for every byte, I'm measuring two pulses (for one 16-bit word)

Writing a sequence of ''0xff 0x00'' bytes (which triggers worst-case behaviour) results in 154 Kwords/sec:
![ff00-conditional](https://user-images.githubusercontent.com/12941778/109559020-720f5a00-7ada-11eb-910c-4cd7eee52cad.jpg)

Writing a sequence of ''0xff 0xff' bytes (which results in best-case behaviour) results in 223 Kwords/sec:
![ffff-conditiional](https://user-images.githubusercontent.com/12941778/109559753-5e182800-7adb-11eb-986c-d7804c28d47b.jpg)

However, unconditionally writing the data raises this to 238 KWords/sec (note the changed timebase):
![ffff-unconditiional](https://user-images.githubusercontent.com/12941778/109559932-9b7cb580-7adb-11eb-95b8-e06f59c9ec5d.jpg)

There is something "magical" about the remaining loop (at least for gcc 9.2.1); if you all but sneeze at the code, the throughput is halved. I godbolted it, and got similar results there. 

![wrx](https://user-images.githubusercontent.com/12941778/109560788-966c3600-7adc-11eb-9987-3ee24b45f91e.jpg)

The WRX pulse is getting a bit narrow at about 150 ns; this is well within the spec (which states > 66 ns) for my STM32F103@72 MHz; a faster MCU ( like the original STM32F407?) might need a delay.

Reply:
I measured 100ns there, so that's still safe.